keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,731 @@
+    // +featureGate=DRAResourceClaimDeviceStatus
+    Devices []AllocatedDeviceStatus `json:"devices,omitempty" protobuf:"bytes,4,opt,name=devices"`
+}
+
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    // Driver specifies the name of the DRA driver whose kubelet
+    // plugin should be invoked to process the allocation once the claim is
+    // needed on a node.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
+    //
+    // +required
+    Driver string `json:"driver" protobuf:"bytes,1,rep,name=driver"`
+
+    // This name together with the driver name and the device name field
+    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
+    //
+    // Must not be longer than 253 characters and may contain one or more
+    // DNS sub-domains separated by slashes.
+    //
+    // +required
+    Pool string `json:"pool" protobuf:"bytes,2,rep,name=pool"`
+
+    // Device references one device instance via its name in the driver's
+    // resource pool. It must be a DNS label.
+    //
+    // +required
+    Device string `json:"device" protobuf:"bytes,3,rep,name=device"`
+
+    // Conditions contains the latest observation of the device's state.
+    // If the device has been configured according to the class and claim
+    // config references, the `Ready` condition should be True.
+    //
+    // +optional
+    // +listType=atomic
+    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,4,opt,name=conditions"`
+
+    // Data contains arbitrary driver-specific data.
+    //
+    // +optional
+    Data runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`
+
+    // NetworkData contains network-related information specific to the device.
+    //
+    // +optional
+    NetworkData NetworkDeviceData `json:"networkData,omitempty" protobuf:"bytes,6,opt,name=networkData"`
+}
+
+// NetworkDeviceData provides network-related details for the allocated device.
+// This information may be filled by drivers or other components to configure
+// or identify the device within a network context.
+type NetworkDeviceData struct {
+    // InterfaceName specifies the name of the network interface associated with
+    // the allocated device. This might be the name of a physical or virtual
+    // network interface.
+    //
+    // +optional
+    InterfaceName string `json:"interfaceName,omitempty" protobuf:"bytes,1,opt,name=interfaceName"`
+
+    // Addresses lists the network addresses assigned to the device's network interface.
+    // This can include both IPv4 and IPv6 addresses.
+    // The addresses are in the CIDR notation, which includes both the address and the
+    // associated subnet mask.
+    // e.g.: "192.0.2.0/24" for IPv4 and "2001:db8::/64" for IPv6.
```

Comment:
So first, those examples are wrong; those are descriptions of subnets, not descriptions of IPs on those subnets. (You need, eg, "192.0.2.5/24").

But also, (1) this is contrary to everything else in kubernetes, where we always use plain IP addresses when referring to IP addresses, and only use CIDR notation for describing subnets and masks; (2) the CIDR notation for interface addresses is only really useful for IPv4 anyway, and doesn't really make sense for IPv6; (3) it is not clear what the user stories would be that would motivate including the link-local-subnet-mask but not any other routing information.

So I think these should be plain IPs.

Reply:
Also we say "IPs" almost everywhere, and where we use "Addresses" it's because it could include non-IP values.  Which is this?

Reply:
"192.0.2.0" and "2001:db8::" _are_ valid IP addresses.  Just try:
```
vm-003 ~ # ip -6 addr add 2001:db8::/64 dev eth0
vm-003 ~ # ip -6 addr show dev eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet6 2001:db8::/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fd00::c0a8:3/120 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::200:ff:fe01:3/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
```
It is a common misunderstanding that host 0 (zero) is reserved. It isn't. But that misunderstanding may be buit-in in older systems/equipment.

Reply:
But because of that misunderstanding 192.0.2.5/24 may be clearer

Reply:
CIDR notation *does* make sense for IPv6. /64 is _not_ required, except for some generated addresses such as the link-local address and SLAAC. IMO it make perfect sense to have *one* /64 address for an entire K8s cluster, and divide it into POD, ClusterIP, and external addresses with CIDR

Reply:
That said, I kind of agree that the mask should be omitted

Reply:
Who is "you" here? The DRA driver? The end user? What is the _exact_ thing that someone can do when the prefix length is exposed here (without needing any other information that is also not exposed here) that they could not do with the prefix length not exposed.

Knowing the subnet size does not tell you how many IPs are _available_ on the subnet so I don't see how it helps with that.

Reply:
The end user can get this information and use it for example to get some hints on who (pod) can talk to whom (pod/external device). The CIDR could also tell you if what you got is what you requested (I requested an IP in the 192.0.2.0/24 subnet, is my address 192.0.2.x/24, or is there something wrong somewhere and my address is 192.0.2.5/8)

This `NetworkDeviceData` struct is likely to evolve in the future when people will start to solve more complex use-cases (as the `NetworkStatus` evolved in the [network-attachment-definition API](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/v1.7.4/pkg/apis/k8s.cni.cncf.io/v1/types.go#L103) (Multus)). So I would prefer to have the addresses in the CIDR format now than just IP addresses and potentially a new field in the future for the CIDR.